File: Projects/Arman/Arman_tools/Transformer.py
""" Transformer 
This class allows for easy traking of 
reference frames. It was developed as
a helper for the Arman project

1/3/2020
Version1.0
"""
import numpy as np

class Transformer:
    def __init__(self, phi, theta, psi, name, protocol=['psi', 'theta', 'phi'], translation=np.array([0.0,0.0,0.0]), parent=None):
        self.name = name
        self.child = None
        self.parent = parent
        self.protocol = protocol
        self.R = np.matrix([[1.0,0.0,0.0],
                           [0.0,1.0,0.0],
                           [0.0,0.0,1.0]])
        self.tf = np.zeros((4,4))
        self.tfI = np.zeros((4,4))
        self.build_transform(phi, theta, psi, translation)
        
    def Rphi(self, phi):
        return np.matrix([[1.0, 0.0, 0.0],
               [0.0, np.cos(phi), np.sin(phi)],
               [0.0, -np.sin(phi), np.cos(phi)]])
        
    def Rtheta(self, theta):
        return np.matrix([[np.cos(theta), 0.0, -np.sin(theta)],
               [0.0, 1.0, 0.0],
               [np.sin(theta), 0.0, np.cos(theta)]])
    
    def Rpsi(self, psi):
        return np.matrix([[np.cos(psi), np.sin(psi), 0.0],
               [-np.sin(psi), np.cos(psi), 0.0],
               [0.0, 0.0, 1.0]])
    
    def build_rotation(self, phi, theta, psi, protocol=None):
        self.R = np.matrix([[1.0,0.0,0.0],
                           [0.0,1.0,0.0],
                           [0.0,0.0,1.0]])
        if protocol is None:
            protocol = self.protocol
        for step in protocol[::-1]:
            if step == 'phi':
                self.R = np.matmul(self.R, self.Rphi(phi))
            elif step == 'theta':
                self.R = np.matmul(self.R, self.Rtheta(theta))
            else:
                self.R = np.matmul(self.R, self.Rpsi(psi))
    
    def add_translation(self, translation, R=None):
        if R is None:
            R = self.R
        n,m = R.shape
        assert len(translation) == n, f'Translation dimension does not equal Rotation dimension: {len(translation)} != {n}'
        new_tf = np.zeros((n+1,m+1))
        for i, row in enumerate(R):
            new_tf[i] = np.concatenate([np.array(row).reshape(m), [translation[i]]])
        new_tf[m,n] = 1.0
        return new_tf

    def build_transform(self, phi, theta, psi, translation):
        self.build_rotation(phi, theta, psi)
        self.tf = self.add_translation(translation, R=self.R)
        self.tfI = self.add_translation(translation, R=self.R.T)
    
    def transform(self, point, inverse=False):
        """ input/output: a 3d cartesian point as numpy array
        """
        if len(point) == 3:
            point = np.concatenate([point, [1.0]])    
        assert point.shape == (4,), 'Illegal point: cannot transform'
        if inverse:
            point = np.array(np.matmul(self.tfI, point.T)).reshape(4) 
            if not self.parent is None:
                return self.parent.transform(point[:3], inverse=True)
        else:
            point = np.array(np.matmul(self.tf, point.T)).reshape(4) 
            if not self.child is None:
                return self.child.transform(point[:3])

        return point[:3]

File: src/dyse-robots/rufous/src/Transformer.cpp
/**
	Transformer class

	Author  : Mitchell Scott (misc4432@colorado.edu)
	Project : Rofous
	Version : 0.1
	Date    : January 31, 2020

**/
#include <math.h>
#include <stdio.h>
#include <ostream>
#include <iostream>
#include <assert.h>
#include <Transformer.hpp>
#define Pi 3.1415926535

using namespace std;

Transformer::Transformer(vector<float> translation, vector<float> reference, vector<char> protocol, string label)
{
	parent = NULL;
	child = NULL;
	name = label;
	build_transform(translation, reference, protocol);
}

Transformer::~Transformer()
{

}

string Transformer::get_name()
{
	return name;
}

void Transformer::set_name(string label)
{
	name = label;
}

void Transformer::Rx(float phi)
{
	vector<vector<float>> rotation = {{1.0, 0.0, 0.0},{0.0, cos(phi), sin(phi)},{0.0, -sin(phi), cos(phi)}};
	Matrix r_mat(rotation);
	R.matmul(&r_mat);
}

void Transformer::Ry(float theta)
{
	vector<vector<float>> rotation = {{cos(theta), 0.0, -sin(theta)},{0.0, 1.0, 0.0},{sin(theta), 0.0, cos(theta)}};
	Matrix r_mat(rotation);
	R.matmul(&r_mat);
}

void Transformer::Rz(float psi)
{
	vector<vector<float>> rotation = {{cos(psi), sin(psi), 0.0},{-sin(psi), cos(psi), 0.0},{0.0, 0.0, 1.0}};
	Matrix r_mat(rotation);
	R.matmul(&r_mat);
}

void Transformer::build_rotation(vector<float> reference, vector<char> protocol)
{
	assert(reference.size() == R.rows());
	for(char step: protocol)
	{
		switch(step){
			case 'x':
				Rx(reference[0]);
				break;
			case 'y':
				Ry(reference[1]);
				break;
			case 'z':
				Rz(reference[2]);
				break;
		}
	}
}

void Transformer::add_translation(vector<float> translation, Matrix* tf)
{
	assert(translation.size() == R.rows());
	vector<float> vect;
	vector<int> js = {0,1,2};
	vector<vector<float>> mat;
	for(int i=0; i < R.rows(); i++)
	{
		vect = R.get(i, js);
		vect.push_back(translation[i]);
		mat.push_back(vect);
		vect.clear();
	}
	mat.push_back({0.0,0.0,0.0,1.0});
	tf->set_matrix(mat);
}

void Transformer::build_transform(vector<float> translation, vector<float> reference, vector<char> protocol)
{
	build_rotation(reference, protocol);
	add_translation(translation, &tf);
	R.T();
	add_translation(translation, &tfI);
	R.T();
}

void Transformer::rotate(Matrix* point, bool inverse)
{
	assert(point->rows() == 3);
	switch(inverse){
		case true:
			R.T();
			point->matmul(&R);
			R.T();
			if(parent)
				parent->rotate(point, true);
			break;
		case false:
			point->matmul(&R);
			if(child)
				child->rotate(point, false);
			break;
	}
}

void Transformer::transform(Matrix* point, bool inverse)
{
	if(point->rows() < 4)
	{
		point->reshape(4,1);
		point->set(3, 1.0);
	}
	switch(inverse){
		case true:
			point->matmul(&tfI);
			if(parent)
				parent->transform(point, true);
			break;
		case false:
			point->matmul(&tf);
			if(child)
				child->transform(point, false);
			break;
	}
	if(point->rows() == 4)
		point->reshape(3,1);
}